Library/Homebrew/formula.rb

```diff
@@ -1625,6 +1625,49 @@ def extract_macho_slice_from(file, arch = Hardware::CPU.arch)
   end
   private :extract_macho_slice_from
 
+  # Generate shell completions for a formula for bash, zsh, and fish, using the formula's binary.
+  #
+  # @param base_name [String] the base name of the generated completion script. Defaults to the formula name.
+  # @param shells [Array<Symbol>] the shells to generate completion scripts for. Defaults to `[:bash, :zsh, :fish]`.
+  # @param binary [Pathname] the binary to use for generating the completion scripts. Defaults to the binary with the
+  # name of the formula.
+  # @param cmd [String] the command to pass to the `binary`. Defaults to 'completion'.
+  # @param shell_as_flag [Boolean] specify if `shells` should each be passed as flags to the `binary`.
+  # Defaults to `false`.
+  sig {
+    params(base_name: String, shells: T::Array[Symbol], binary: Pathname, cmd: String,
+           shell_prefix: T.nilable(T.any(Symbol, String))).void
+  }
+  def generate_completions_from_executable(base_name: name,
+                                           shells: [:bash, :zsh, :fish],
+                                           executable: bin/base_name,
+                                           cmd: "completion",
+                                           shell_parameter_format: nil)
+    completion_script_path_map = {
+      bash: bash_completion/base_name,
+      zsh:  zsh_completion/"_#{base_name}",
+      fish: fish_completion/"#{base_name}.fish",
+    }
+
+    shells.each do |shell|
+      script_path = completion_script_path_map[shell]
+      shell_parameter = if shell_parameter_format.nil?
+        shell.to_s
+      elsif shell_parameter_format == :flag
+        "--#{shell}"
+      elsif shell_parameter_format == :arg
+        "--shell=#{shell}"
+      elsif shell_parameter_format == :none
+        nil
+      else
+        "#{shell_parameter_format}#{shell}"
```

Comment:
Which formulae did you have in mind for this case? I wonder if we should handle it at all. Or, if we will, let `shell_parameter_format` be a method (or a `Proc` object?) that takes an argument and spits out the right format.

Reply:
For example, the `certigo` formula currently does this.
```ruby
# Install bash completion
output = Utils.safe_popen_read("#{bin}/certigo", "--completion-script-bash")
(bash_completion/"certigo").write output

# Install zsh completion
output = Utils.safe_popen_read("#{bin}/certigo", "--completion-script-zsh")
(zsh_completion/"_certigo").write output
```

This would turn into 
```ruby
generate_completions_from_executable(cmd: "", shells: [:bash, :zsh], shell_parameter_format: "--completion-script-")
``` 
and would not work without the current handling. 

Another example is the `earthly` formula.
```ruby
bash_output = Utils.safe_popen_read("#{bin}/earthly", "bootstrap", "--source", "bash")
(bash_completion/"earthly").write bash_output
zsh_output = Utils.safe_popen_read("#{bin}/earthly", "bootstrap", "--source", "zsh")
(zsh_completion/"_earthly").write zsh_output

generate_completions_from_executable(cmd: "bootstrap", shells: [:bash, :zsh], shell_parameter_format: "--source ")
# or 
generate_completions_from_executable(cmd: "bootstrap", shells: [:bash, :zsh], shell_parameter_format: "--source=")
```

Reply:
In which case it might be useful to let `shell_parameter_format` be some sort of callable so we can accommodate formulae like `rabbitmq`:
```ruby
    (bash_completion/"rabbitmqadmin.bash").write Utils.safe_popen_read("#{sbin}/rabbitmqadmin", "--bash-completion")
```
Maybe `yield` will do enough here. But feel free not to handle this if it's too complicated.

Reply:
Another option might be to use a placeholder like `@@SHELL@@` (maybe that's too long) so we can do (in addition:

```ruby
    generate_completions_from_executable(base_name: "rabbitmqadmin", shell_parameter: "--@@SHELL@@-completion")
```

If we use a block, I think it might be smart to have the block set the list of arguments, rather than just the shell parameter. It feels more flexible, that way. Something like this:

```ruby
    generate_completions_from_executable do |shell|
      [bin/"rabbitmqadmit", "--#{shell}-completion"]
    end
```

Reply:
I like your suggestions, but I feel like this extension to cover even more formulae with the new syntax might warrant a followup PR.

Either someone else can take this up, or I might in the future when I familiarized myself a bit more with blocks/`Proc`s 😄